skyportal/models.py

```diff
@@ -2124,28 +2110,77 @@ class Annotation(Base):
             'annotation with multiple fields from each origin is allowed.'
         ),
     )
-    obj_id = sa.Column(
-        sa.ForeignKey('objs.id', ondelete='CASCADE'),
-        nullable=False,
-        index=True,
-        doc="ID of the Annotation's Obj.",
-    )
 
-    obj = relationship('Obj', back_populates='annotations', doc="The Annotation's Obj.")
-    groups = relationship(
-        "Group",
-        secondary="group_annotations",
-        cascade="save-update, merge, refresh-expire, expunge",
-        passive_deletes=True,
-        doc="Groups that can see the annotation.",
-    )
+    @classmethod
+    def backref_name(cls):
+        if cls.__name__ == 'Annotation':
+            return "annotations"
+        if cls.__name__ == 'AnnotationOnSpectrum':
+            return 'annotations_on_spectra'
+
+    @declared_attr
+    def author(cls):
+        return relationship(
+            "User",
+            back_populates=cls.backref_name(),
+            doc="Annotation's author.",
+            uselist=False,
+            foreign_keys=f"{cls.__name__}.author_id",
+        )
+
+    @declared_attr
+    def author_id(cls):
+        return sa.Column(
+            sa.ForeignKey('users.id', ondelete='CASCADE'),
+            nullable=False,
+            index=True,
+            doc="ID of the Annotation author's User instance.",
+        )
+
+    @declared_attr
+    def obj_id(cls):
+        return sa.Column(
+            sa.ForeignKey('objs.id', ondelete='CASCADE'),
+            nullable=False,
+            index=True,
+            doc="ID of the Annotation's Obj.",
+        )
+
+    @declared_attr
+    def obj(cls):
+        return relationship(
+            'Obj',
+            back_populates=cls.backref_name(),
+            doc="The Annotation's Obj.",
+        )
+
+    @declared_attr
+    def groups(cls):
+        return relationship(
+            "Group",
+            secondary="group_" + cls.backref_name(),
+            cascade="save-update, merge, refresh-expire, expunge",
+            passive_deletes=True,
+            doc="Groups that can see the annotation.",
+        )
 
     def construct_author_info_dict(self):
         return {
             field: getattr(self.author, field)
             for field in ('username', 'first_name', 'last_name', 'gravatar_url')
         }
 
+
+class Annotation(Base, AnnotationMixin):
+    """A sortable/searchable Annotation made by a filter or other robot,
```

Comment:
Perhaps note that this is explicitly for annotations on sources in this docstring